lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -0,0 +1,249 @@
+    """For a given PID file name keep track of discovered PIDs and UUIDs as
+    Tool Meister directories are `load()`ed.  The `killem()` method is invoked
+    by the caller at their discretion.
+
+    The `killem()` method clears out all accumlated data.
+    """
+
+    def __init__(self, file_name: str, display_name: str):
+        self.file_name = file_name
+        self.display_name = display_name
+        self.pids_by_uuid = {}
+        self.uuid_to_tmdir = {}
+
+    def load(self, tm_dir: pathlib.Path, uuid: str) -> bool:
+        """Load a PID from the given directory associated with the given UUID.
+
+        Records the loaded PID if it has a live process associated with it and
+        returns True, otherwise returns False.
+        """
+        try:
+            pid = (tm_dir / self.file_name).read_text()
+        except FileNotFoundError:
+            return False
+        try:
+            self.pids_by_uuid[uuid] = psutil.Process(pid)
+        except psutil.NoSuchProcess:
+            return False
+        self.uuid_to_tmdir[uuid] = tm_dir
+        return True
+
+    def killem(self, echo: Callable[[str], None]) -> None:
+        """Kill all PIDs found, and their children."""
+        if not self.pids_by_uuid:
+            return
+        echo(f"Killing {self.display_name} PIDs ...")
+        pids_by_uuid, self.pids_by_uuid = self.pids_by_uuid, {}
+        uuid_to_tmdir, self.uuid_to_tmdir = self.uuid_to_tmdir, {}
+        for uuid, proc in pids_by_uuid.items():
+            pid = proc.pid
+            echo(f"\tKilling {pid} (from {uuid_to_tmdir[uuid]})")
+            try:
+                kill_family(proc)
+            except Exception as exc:
+                echo(f"\t\terror killing {pid}: {exc}", err=True)
+
+
+def gen_run_directories(pbench_run: pathlib.Path) -> Tuple[pathlib.Path, str]:
+    """Generate the list of run directories available under ${pbench_run},
+    yielding a Path object for that directory, along with its recorded
+    UUID.
+
+    Yields a tuple of the run directory Path object and associated UUID.
+    """
+    for entry in pbench_run.iterdir():
+        if not entry.is_dir():
+            continue
+        tm_dir = entry / "tm"
+        try:
+            uuid = (tm_dir / ".uuid").read_text()
+        except FileNotFoundError:
+            # This is either not a pbench run directory, or the Tool Meister
+            # sub-system was not orchestrated by pbench-tool-meister-start
+            # for this run.
+            continue
+        yield tm_dir, uuid
+
+
+def gen_host_names(tm_dir: pathlib.Path) -> str:
+    """Read the registered tool data saved for this run and return the list
+    of remote hosts.
+    """
+    run_dir = tm_dir.parent
+    tool_group_dirs = list(run_dir.glob(f"{ToolGroup.TOOL_GROUP_PREFIX}-*"))
+    if not tool_group_dirs:
+        return
+
+    lrh = LocalRemoteHost()
+
+    for tool_group_dir in tool_group_dirs:
+        tg_name = tool_group_dir.name.split("-", 2)[2]
```

Comment:
A comment describing the semantics of your `<x>-<y>-<tool-group>` pattern would be a help here! `tools-v1-<name>` may not be obvious to casual browsers here...

Reply:
Perhaps it would be better to have a `staticmethod` of the ToolsGroup class?

Reply:
Perhaps it is best to just keep it this way for now, and rely on PR #3038 to correct all this?